Add nRF52840 USB device controller driver

sundar-subramaniyan opened the pull request:
This pull request is created to basically review the nRF52840 USB device controller driver.

carlescufi commented:
CC @pawelzadrozniak

carlescufi commented:
@cvinayak can you take a look at the clock control driver changes?

cvinayak commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -12,10 +12,16 @@
 #include <clock_control.h>
 #include <misc/__assert.h>
 #include <arch/arm/cortex_m/cmsis.h>
+#include <nrf_power.h>
```
Have a separate commit for  nrf_power_clock changes.

sundar-subramaniyan replied:
Okay, will do.

cvinayak commented:
@sundar-subramaniyan hci_usb uses ~7392 bytes more ROM than hci_uart, but uses 1084 bytes less RAM, can you justify?

cvinayak commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -253,7 +260,12 @@ static void _power_clock_isr(void *arg)
 	ctto = (NRF_CLOCK->EVENTS_CTTO != 0);
 
-	__ASSERT_NO_MSG(pof || hf || hf_intenset || lf || done || ctto);
+	usb_detected = (NRF_POWER->EVENTS_USBDETECTED != 0);
```
conditional compile under CONFIG_USB.

sundar-subramaniyan replied:
Yes, I had forgotten this one. Will do. Although CONFIG_USB should do the job, shouldn't it be CONFIG_USB_NRF52 rather?

cvinayak replied:
Maybe as below (tho for now only nRF52840 SoC will be supported)

```
#if defined(CONFIG_USB)
#if defined(CONFIG_SOC_NRF52840)
....
#endif
#endif
```

ydamigos commented:
Hi guys,
now that you are introducing nRF52 usb driver, could you have a look at PR #4376, which add USB bindings for DT. Does these bindings work for nRF52 usb driver?

SebastianBoe commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -12,10 +12,16 @@
 #include <clock_control.h>
 #include <misc/__assert.h>
 #include <arch/arm/cortex_m/cmsis.h>
+#include <nrf_power.h>
```
Hi,

the path "#include <nrfx/hal/nrf_power.h>" is valid because the CMakeLists.txt code has been written like this:

```
  zephyr_include_directories(nrfx)
  zephyr_include_directories(nrfx/drivers/include)
  zephyr_include_directories(nrfx/hal)
  zephyr_include_directories(nrfx/mdk)
  zephyr_include_directories(.)

```

I will leave it to @anangl to comment on what path is desirable, but for this PR I would assume just <nrf_power.h> makes sense as that is consistent with the other users of the nrfx hal, and support for it was explicitly given when this line of code was added:

`  zephyr_include_directories(nrfx/hal)
`

```
sebo@mach:~/zephyr$ ack "<nrf_"
drivers/usb/device/usb_dc_nrf5.c
22:#include <nrf_clock.h>
23:#include <nrf_power.h>
30:#include <nrf_usbd.h>

drivers/gpio/gpio_nrf5.c
18:#include <nrf_gpiote.h>

drivers/entropy/entropy_nrf5.c
10:#include <nrf_rng.h>

drivers/clock_control/nrf5_power_clock.c
15:#include <nrf_power.h>

arch/arm/soc/nordic_nrf5/nrf52/power.c
12:#include <nrf_power.h>

ext/hal/nordic/nrfx/hal/nrf_ecb.c
35:#include <nrf_ecb.h>

ext/hal/nordic/nrfx/hal/nrf_nvmc.c
37:#include <nrf_nvmc.h>

ext/hal/nordic/nrfx/drivers/nrfx_common.h
40:#include <nrf_peripherals.h>

```

sundar-subramaniyan replied:
I'll wait for anangl's comments for this one.

sundar-subramaniyan commented on arch/arm/soc/nordic_nrf5/include/nrf5_common.h:
```diff
@@ -61,6 +61,13 @@
 #define NRF52_IRQ_I2S_IRQn                     37
 #define NRF52_IRQ_FPU_IRQn                     38
 
+/**
```
Okay. Will do.

sundar-subramaniyan commented on arch/arm/soc/nordic_nrf5/include/nrf5_common.h:
```diff
@@ -61,6 +61,13 @@
+ *
+ * The interrupt numbers below are specific to only NRF52840 SoC
+ */
+#define NRF52840_IRQ_USBD_IRQn                 39
```
I thought this would be confusing and went with NRF52840. But what you say about future chipsets makes sense. I'll change it to just NRF52 and remove the comments too.

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -12,10 +12,16 @@
 static u8_t m16src_ref;
 static u8_t m16src_grd;
 static u8_t k32src_initialized;
+static struct nrf5_usb_event_cb usb_event;
```
I actually had a private pointer (void * to have struct nrf_usbd_ctx *ctx) member in the structure earlier and removed it later. I'll keep it as a simple function pointer. I am managing now using the global usbd context.

sundar-subramaniyan replied:
Alright.

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -239,6 +245,7 @@ static int _k32src_start(struct device *dev, clock_control_subsys_t sub_system)
 static void _power_clock_isr(void *arg)
 {
 	u8_t pof, hf_intenset, hf_stat, hf, lf, done, ctto;
+	u8_t usb_detected, usb_power_ready, usb_removed;
```
Okay. How about usb_detected and usb_removed? Should I rename them to "usb_det" and "usb_rem" respectively?

cvinayak replied:
I just want to reflect the event register names, but I have no strong feeling on the naming, just pick or do nothing on this comment.

sundar-subramaniyan replied:
Okay. I'll have the same names as event register names.

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -310,6 +322,24 @@ static void _power_clock_isr(void *arg)
+	if (usb_detected) {
+		NRF_POWER->EVENTS_USBDETECTED = 0;
+		if (usb_event.cb)
+			usb_event.cb(NRF_POWER_EVENT_USBDETECTED);
```
Sure. Will do.

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -329,6 +359,22 @@ static int _clock_control_init(struct device *dev)
 	return 0;
 }
 
+void nrf5_power_clock_install_usb_cb(void (*cb)(nrf_power_event_t ev))
```
This is what I wanted to discuss. Basically these are registered at run-time now. I don't see a use-case for run-time registration and de-registration actually. But suppose, if the USB driver isn't enabled and before that the USB events happens to be triggered we'd miss them. So the USB cable needs to be unplugged and re-plugged to kick-start enumeration, unless the power/clock interrupt enabling part is done from usb_enable().

So if I want to keep the function prototypes, I may have to add a new header file or alternatively if there are some suitable device I can initialize with DEVICE_INIT() I can give it a try. But so far, some standard interface is available for clock control and gpio toggling. I'll give another look and discuss with you.

cvinayak replied:
- Isnt there a status register in USB device controller, that can be checked on late registration?
- Zephyr driver model (most) use static compile/link time symbol resolution, atleast for driver bindings.

Let us, for now, try the DEVICE_INIT way of static assignment, which is run-time efficient (in ISR).

sundar-subramaniyan replied:
I see USBREGSTATUS under power supply which provides the USB detected and powered conditions. I'll have to read them out in the USB driver. I can either read it in the USB driver or have it part of nrf5_power_clock.c file together with the other changes. What would you recommend?

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -329,6 +359,22 @@ static int _clock_control_init(struct device *dev)
+void nrf5_power_clock_install_usb_cb(void (*cb)(nrf_power_event_t ev))
+{
+	usb_event.cb = cb;
+	NRF_POWER->INTENSET |= (NRF_POWER_INT_USBDETECTED_MASK |
```
As mentioned in the previous comment, run-time registration seems to be required. I'll have register/unregister interface in that case.

sundar-subramaniyan commented on drivers/usb/device/CMakeLists.txt:
```diff
@@ -1,2 +1,3 @@
 zephyr_sources_ifdef(CONFIG_USB_DW       usb_dc_dw.c)
 zephyr_sources_ifdef(CONFIG_USB_DC_STM32 usb_dc_stm32.c)
+zephyr_sources_ifdef(CONFIG_USB_NRF52840 usb_dc_nrf5.c)
```
Short form for device controller. I just wanted to stick to the naming convention used already by other drivers.

finikorg replied:
lgtm

sundar-subramaniyan commented on drivers/usb/device/Kconfig:
```diff
@@ -62,6 +62,59 @@ config USB_DC_STM32_RAM_SIZE
+	bool
+	prompt "Nordic Semiconductor nRF52840 USB Device Controller Driver"
+	depends on SOC_SERIES_NRF52X
+	depends on BOARD_NRF52840_PCA10056
```
Alright. Will do.

sundar-subramaniyan commented on drivers/usb/device/Kconfig:
```diff
@@ -62,6 +62,59 @@ config USB_DC_STM32_RAM_SIZE
+	depends on SOC_SERIES_NRF52X
+	depends on BOARD_NRF52840_PCA10056
+	select USB_DEVICE_DRIVER
+	default n
```
Sure. Will remove it.

sundar-subramaniyan replied:
I wanted to keep the "default n" since it improves readability. Please suggest if you think otherwise.

finikorg replied:
It is OK to have `default n` since all other drivers have it.

sundar-subramaniyan commented on drivers/usb/device/Kconfig:
```diff
@@ -62,6 +62,59 @@ config USB_DC_STM32_RAM_SIZE
+config USBD_NRF52840_IRQ_PRI
+	int
+	prompt "USBD Interrupt priority"
+	depends on USB_NRF52840
```
Okay, will do.

sundar-subramaniyan commented:
@cvinayak Do you mean hci_usb including the usb driver? The usb driver right now allocates memory pool for all 18 endpoints which might not be required for hci_usb. We can ideally configure just the control and bulk/interrupt endpoints. The ISO endpoints consume most memory (1024 bytes each).

sundar-subramaniyan commented on samples/bluetooth/hci_usb/prj.conf:
```diff
@@ -6,3 +6,8 @@ CONFIG_SERIAL=y
 CONFIG_UART_INTERRUPT_DRIVEN=y
 CONFIG_BT=y
 CONFIG_BT_HCI_RAW=y
+CONFIG_USB_NRF52840=y
```
Yes, right. I put this just for others to compile and try it out. Mostly this commit won't be in the review as I hope the usb_transfer API addition will fix this problem. I'll remove this if I plan to keep the main.c file changes for review in the next review.

sundar-subramaniyan commented on samples/bluetooth/hci_usb/src/main.c:
```diff
@@ -661,14 +661,26 @@ DEVICE_INIT(btusb, "btusb", &btusb_init,
 			break;
 		/* TODO: Handle other error codes */
 		default:
+			if (wrote < len) {
```
Okay. Actually this is to get the bt-usb working as it's only a hack. The wrote pointer is always written with the value in usb_write() handler in the usb driver so I left it un-initialized. I guess the correct way to fix would be to modify the usb_write() to usb_transfer() in synchronous mode (once the usb_transfer API change is merged) which takes care of all these handling properly .

carlescufi commented:
@sundar-subramaniyan CI is failing because you are using nRF52840 registers in clock_control without and #ifdef, so the build for nRF51 and nRF52832 fails.

sundar-subramaniyan commented:
@carlescufi I'll address it along with the other review comments.

cvinayak commented:
@sundar-subramaniyan you can review https://github.com/zephyrproject-rtos/zephyr/pull/6243 for inspirations and vice versa.

galak reviewed (dismissed):
Needs device tree support, look at PR #4376 for example

sundar-subramaniyan commented:
@cvinayak Sure, I'll take a look at it.

sundar-subramaniyan commented:
@galak Okay, I'll look at the DT change and rework on the PR.

sundar-subramaniyan committed:
dts: bindings: Add yaml file for nRF52840 USBD support to DT

Add yaml files to DT to support nRF52840 USBD

Origin: Original

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan committed:
dts: arm: nrf52840_pca10056: Use DT to configure USB

Configure USB using DT for nrf52840_pca10056

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan committed:
clock_control: nrf5: Add USBD power event support for USB enumeration

nRF52840 USBD requires power events to start USB enumeration.

Add APIs to enable/disable the power events, read the USB VBUS detect
and output power ready statuses which can then be used by the USB driver
in the enumeration sequence.

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan committed:
drivers: usbd_nrf5: Add nRF52840 USB Device Controller Driver

* Add USB device controller driver for nRF52840 SoC.
* Supports only Bulk and Interrupt endpoints for now.
* ISO endpoint support is WIP.

What's tested:
* BTUSB - LE Scanning and connection.

Known issues in btusb tests:
 * Observed momentary delays during LE scanning. This is possibly
due to hci_usb's way of calling the usb_write().

 * Sometimes, when LE scan stop command is issued, response is not
sent and as a result USB stops functioning. This is also related
to the way btusb is handling the write(). For some reason, the
btusb class handler seems to issue another data packet write before
the USB control transaction is complete.

Miscellaneous:
* Lots of TODO items to discuss and fix.

Origin: Original

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan commented:
@cvinayak @galak I have reworked and addressed the changes requested. Please have a look.

sundar-subramaniyan commented:
@carlescufi I have added the necessary ifdef checks and hopefully CI should pass this time.

ydamigos commented on drivers/usb/device/Kconfig:
```diff
@@ -39,6 +39,39 @@ config USB_DC_STM32
 	help
 	  STM32 family USB device controller Driver.
 
+config USB_DC_STM32_IRQ_PRI
```
Please remove this.

sundar-subramaniyan replied:
My mistake. I'll fix this and update it.

ydamigos commented on drivers/usb/device/Kconfig:
```diff
@@ -39,6 +39,39 @@ config USB_DC_STM32
+	help
+	  USB interrupt priority.
+
+config USB_DC_STM32_EP_NUM
```
Please remove this.

sundar-subramaniyan replied:
My mistake. I'll fix this and update it.

ydamigos commented on drivers/usb/device/Kconfig:
```diff
@@ -39,6 +39,39 @@ config USB_DC_STM32
+	help
+	  Number of supported endpoints (including ep0).
+
+config USB_DC_STM32_RAM_SIZE
```
Please remove this.

sundar-subramaniyan replied:
My mistake. I'll fix this and update it.

ydamigos reviewed (commented):
During a rebase some old STM32 config options were added.

sundar-subramaniyan commented:
@ydamigos 

> During a rebase some old STM32 config options were added.

Thanks for pointing it out. I have fixed them now.

cvinayak commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -352,3 +394,73 @@ DEVICE_AND_API_INIT(clock_nrf5_k32src,
+
+static bool nrf5_power_usbregstatus_outrdy_get(void)
+{
+	return nrf_power_usbregstatus_outrdy_get();
```
What is this nrf_power namespace?

sundar-subramaniyan replied:
Those are the HAL functions from ext. I'm using them instead of re-implementing again.

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -236,9 +240,21 @@ static int _k32src_start(struct device *dev, clock_control_subsys_t sub_system)
 }
 
+#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+static void nrf5_usbd_power_event_deliver(nrf_power_event_t event)
```
Okay. Will do.

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -352,3 +394,73 @@ DEVICE_AND_API_INIT(clock_nrf5_k32src,
+	       NRF_POWER_INT_USBPWRRDY_MASK;
+
+	if (enable) {
+		NRF_POWER->INTENSET |= mask;
```
There are other interrupts in NRF_POWER like SLEEPENTER, SLEEPEXIT which may be configured in future so we need to OR with the previously set interrupts right? Or is the INTENSET register internally ORs with the already set interrupt mask?

pawelzadrozniak replied:
Writing to INTENSET (-CLR) enables (disables) the insterrupts only, so result of INTENSET = mask is (almost) equivalent to INTEN |= mask. The difference is that using INENSET is an atomic operation.
nrf_power_int_enable() form nrf_power.h may be used as well.

sundar-subramaniyan replied:
I'm all for using the nrfx HAL implementation. From the current code perspective, I did not want to mix the direct register read/write and nrfx together. Well if new code should use nrfx instead we can go with that.
@cvinayak What is your opinion on this?

sundar-subramaniyan commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -352,3 +394,73 @@ DEVICE_AND_API_INIT(clock_nrf5_k32src,
+	if (enable) {
+		NRF_POWER->INTENSET |= mask;
+	} else {
+		NRF_POWER->INTENSET &= ~mask;
```
My bad. It should be INTENCLR. I'll fix it.

pawelzadrozniak commented on drivers/clock_control/nrf5_power_clock.c:
```diff
@@ -253,7 +269,16 @@ static void _power_clock_isr(void *arg)
 	ctto = (NRF_CLOCK->EVENTS_CTTO != 0);
 
+#if defined(CONFIG_USB) && defined(CONFIG_SOC_NRF52840)
+	usb_detected = (NRF_POWER->EVENTS_USBDETECTED != 0);
```
nrf_power.h (and other headers from ext/nordic/nrfx/hal) provides inline functions for handling register level operations in more readable way, i.e. nrf_power_event_check/clear.
It's OK for now; the target is to use nrfx drivers in most of shims.

sundar-subramaniyan replied:
We can use nrfx if it's okay to mix with direct register read/write.
@cvinayak What do you think?

cvinayak replied:
Yes, please use nrfx interfaces, at least the inline ones should not be of any overhead.

sundar-subramaniyan replied:
Okay.

sundar-subramaniyan commented:
@cvinayak @pawelzadrozniak I have addressed the comments in clock_control commit.

carlescufi commented:
@sundar-subramaniyan can we split out the hci_usb patch into another PR and finish the discussion regarding that there?

sundar-subramaniyan commented:
@carlescufi I have split the PR now. This PR now has only the USB driver changes and a new PR #6468 is created for the hci_usb changes.

carlescufi commented:
@cvinayak @galak mind re-reviewing this?

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+ * Find alternatives for SET_BIT and friends
+ * For now, define them here
+ */
+#define SET_BIT(W, B)  ((W) |= (u32_t)(1U << (B)))
```
Use the `BIT` and `BIT_MASK` macros directly, I don't think there's a need for these 3 extra macros here. See https://github.com/zephyrproject-rtos/zephyr/blob/master/include/misc/util.h#L118

sundar-subramaniyan replied:
Okay, will do.

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+	return &usbd_ctx;
+}
+
+static inline bool nrf5_usbd_ep_is_valid(const u8_t ep)
```
In general for all static functions: remove all `nrf5_usbd_` prefix, this is not needed at all and makes the code too verbose

sundar-subramaniyan replied:
Okay. Will do.

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+	clock = device_get_binding(CONFIG_CLOCK_CONTROL_NRF5_M16SRC_DRV_NAME);
+	if (!clock) {
+		SYS_LOG_ERR("NRF5 HF Clock device not found!");
+		goto err_clk_dev;
```
`return ret` here directly, no need for `goto`

sundar-subramaniyan replied:
Ack.

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+	if (ret) {
+		SYS_LOG_ERR("NRF5 HF clock %s fail: %d",
+			    on ? "start" : "stop", ret);
+		goto err_clk_start_stop;
```
`return ret` here directly, no need for `goto`

sundar-subramaniyan replied:
Ack.

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+
+	SYS_LOG_DBG("HF clock %s success", on ? "start" : "stop");
+
+err_clk_dev:
```
remove those 2 labels

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+
+	ret = nrf5_usbd_power_int_enable(true);
+	if (ret) {
+		goto err_pwr_int_enable;
```
`return ret` here directly, no need for `goto`

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+	nrf5_usbd_disable_interrupts();
+	nrf5_usbd_power_int_enable(false);
+
+err_pwr_int_enable:
```
remove those 2 labels

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+
+	ret = nrf5_usbd_hf_clock_enable(false, false);
+	if (ret) {
+		goto err_clk_disable;
```
`return ret` here directly, no need for `goto`

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+
+	ret = nrf5_usbd_power_int_enable(false);
+	if (ret) {
+		goto err_pwr_int_disable;
```
`return ret` here directly, no need for `goto`

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+
+	ctx->attached = false;
+
+err_clk_disable:
```
remove those 2 labels

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+		if (read_bytes) {
+			*read_bytes = ep_ctx->buf.len;
+		}
+		goto done;
```
`return ret` here directly, no need for `goto`

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2256 @@
+		*read_bytes = bytes_to_copy;
+	}
+
+done:
```
remove label

sundar-subramaniyan replied:
Ack

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2223 @@
+	}
+
+	ret = usb_dc_ep_read_continue(ep);
+	if (ret) {
```
remove if statement and return ret directly

cvinayak reviewed (approved):
clock_control changes LGTM

IvanSanchez commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2223 @@
+		switch (ep_ctx->cfg.type) {
+		case USB_DC_EP_CONTROL:
+			nrf_usbd_epout_clear(ep);
+			break;
```
Spurious `break` that could be removed - is there any difference between `USB_DC_EP_CONTROL` and the other two cases?

carlescufi commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2223 @@
+/** Interrupt mask to event register map */
+#define USBD_INT_CNT 26
+
+static u32_t event_map[USBD_INT_CNT] = {
```
const?

IvanSanchez commented:
I have sucessfully compiled some of the USB examples with this driver, but the nRF52840 board definition needs to be updated. @sundar-subramaniyan , please have a look at https://github.com/sundar-subramaniyan/zephyr/pull/1

I'll be testing the USB examples from the USB host side of things.

carlescufi commented:
@sundar-subramaniyan can you fix the conflicts?
@finikorg are you OK with this being merged?

sundar-subramaniyan commented:
@carlescufi Rebased and fixed the conflicts.

sundar-subramaniyan commented:
@IvanSanchez I looked at your PR. It looks okay, however I think this driver is kind of in experimental phase and I think USB should be enabled by default after having tested all the examples first. Secondly, there's also applications that doesn't require USB and keeping it default will consume memory unnecessarily. Should we enable this rather in some new prj.conf files under USB based samples to start with? Later, once the driver is stable we can add it in the defconfig.

@carlescufi What do you think?

sundar-subramaniyan commented:
@IvanSanchez Had a little chat with Carles and he feels it's okay to have the default as yes in the defconfig as long as it is enclosed by "USB" like you have it now. Would you like to create a PR for that? Maybe after this one merges?

IvanSanchez commented:
@sundar-subramaniyan You can just merge https://github.com/sundar-subramaniyan/zephyr/pull/1 - my changes will go into *your* development branch, and will automatically become part of this PR.

finikorg reviewed (commented):
LGTM, not tested yet.
PS: you may want to select this driver if USB enabled the same way we have for quark_se: in `arch/x86/soc/intel_quark/quark_se/Kconfig.defconfig.series`

```
if USB
config USB_DW
	def_bool y
config GPIO
	def_bool y
config USB_DW_IRQ_PRI
	default 3
endif # USB
```

You should also add sanitycheck build configuration the same way we have for other boards.

carlescufi commented:
@sundar-subramaniyan ok before i merge this could you do 2 things:

1) Incorporate @IvanSanchez's changes 
2) Enable a configuration for sanitycheck. @finikorg I see no USB in `tests/`, where is the sanitycheck build config that you mention?

finikorg commented:
@carlescufi sanity check configurations can be found in `samples/subsys/usb/`, like `samples/subsys/usb/hid/sample.yaml`. You can run tests with `sanitycheck -t usb`.

sundar-subramaniyan commented:
@IvanSanchez Alright, I'll pull your change and update this PR.

sundar-subramaniyan commented:
@carlescufi Sure, will do both the changes and update the PR.

sundar-subramaniyan commented:
@finikorg I'll add the defconfig and sanity check configurations. Thanks for the pointers.

sundar-subramaniyan committed:
boards: arm: nrf: Make USB default for nrf52840_pca10056 board

Make USB as default for the nrf52840_pca10056 board.

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan commented:
@carlescufi Updated the PR with the defconfig and sanity check commits.

sundar-subramaniyan commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2213 @@
+#include <nrf_usbd.h>
+
+#define SYS_LOG_LEVEL CONFIG_SYS_LOG_USB_DRIVER_LEVEL
+#define SYS_LOG_DOMAIN "USBD"
```
Okay. Will do.

sundar-subramaniyan commented on drivers/usb/device/usb_dc_nrf5.c:
```diff
@@ -0,0 +1,2213 @@
+		  FIFO_ELEM_COUNT, FIFO_ELEM_ALIGN);
+
+/**
+ * @brief Endpoint buffer pool
```
L231

sundar-subramaniyan commented:
@jfischer-phytec-iot I have addressed some of your comments and updated the PR.

sundar-subramaniyan commented:
@carlescufi There's a sticky situation with the sanity checks. One of the test cases related to ieee802154 fails to compile for nrf52840_pca10056. It looks like after the addition of usb_device to the supported list, the script compiles the sample net/wpanusb and it fails at an undefined reference for symbol cc2520_configure_gpios().

First of all cc2520 shouldn't be compiled at all for nrf52840 since it has its own ieee802154 radio.
The prj.conf simply hard codes the CONFIG_IEEE802154_CC2520=y and that seems to be the issue.
I see there's a platform_exclude to instruct the script to avoid the build for nrf52840_pca10056 but that's not the correct way to fix it.

sundar-subramaniyan committed:
tests: samples: Add nrf52840_pca10056 board to USB test cases

Add nrf52840_pca10056 to platform whitelist for hci_usb and cdc_acm
samples.

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan committed:
tests: samples: Exclude nrf52840_pca10056 from wpanusb sample

With usb_device support added to nrf52840_pca10056 board, sanity
check script compiles the sample net/wpanusb and it fails with
undefined reference for symbol cc2520_configure_gpios.

First of all cc2520 shouldn't be compiled at all for nrf52840
since it has its own ieee802154 radio. The prj.conf under the
sample samples/net/wpanusb simply hard codes the config
CONFIG_IEEE802154_CC2520=y and is causing the ieee80211_cc2520.c
file to be compiled. If it's removed from the configuration file
we get other errors. So it is not straight forward.

For now, as a workaround, exclude the nrf52840_pca10056 platform
until there's a proper fix.

Signed-off-by: Sundar Subramaniyan <sundar.subramaniyan@gmail.com>

sundar-subramaniyan commented:
@carlescufi I have added another commit to workaround the sanity check failure. Please see if we can live with that.

sundar-subramaniyan commented:
@carlescufi Seems like the sanity check bypass for ieee802154 has worked out. If we want this to be merged, I'll have to keep it before the sanity check addition commit so that each commit compiles atomically. Please have a look.

carlescufi commented:
Tested with the `hci_usb` sample. There are issues with other classes but we will figure them out once merged.
